diff -uNr iptables-1.2.9_20040304.orig/include/libippool/ip_pool_support.h iptables-1.2.9_20040304/include/libippool/ip_pool_support.h
--- iptables-1.2.9_20040304.orig/include/libippool/ip_pool_support.h	2000-08-27 09:47:46.000000000 +0200
+++ iptables-1.2.9_20040304/include/libippool/ip_pool_support.h	2004-03-04 16:24:52.757048352 +0100
@@ -2,7 +2,7 @@
 #ifndef _IP_POOL_SUPPORT_H
 #define _IP_POOL_SUPPORT_H
 
-#include <iptables.h>
+#include <iptables/iptables.h>
 #include <linux/netfilter_ipv4/ip_conntrack.h>
 #include <linux/netfilter_ipv4/ip_pool.h>
 
diff -uNr iptables-1.2.9_20040304.orig/include/libiptc/libip6tc.h iptables-1.2.9_20040304/include/libiptc/libip6tc.h
--- iptables-1.2.9_20040304.orig/include/libiptc/libip6tc.h	2003-05-05 21:33:40.000000000 +0200
+++ iptables-1.2.9_20040304/include/libiptc/libip6tc.h	2004-03-04 16:24:52.763047440 +0100
@@ -2,7 +2,7 @@
 #define _LIBIP6TC_H
 /* Library which manipulates firewall rules. Version 0.2. */
 
-#include <libiptc/ipt_kernel_headers.h>
+#include <iptables/libiptc/ipt_kernel_headers.h>
 #include <linux/netfilter_ipv6/ip6_tables.h>
 
 #ifndef IP6T_MIN_ALIGN
diff -uNr iptables-1.2.9_20040304.orig/include/libiptc/libiptc.h iptables-1.2.9_20040304/include/libiptc/libiptc.h
--- iptables-1.2.9_20040304.orig/include/libiptc/libiptc.h	2003-05-02 17:30:11.000000000 +0200
+++ iptables-1.2.9_20040304/include/libiptc/libiptc.h	2004-03-04 16:24:52.765047136 +0100
@@ -2,7 +2,7 @@
 #define _LIBIPTC_H
 /* Library which manipulates filtering rules. */
 
-#include <libiptc/ipt_kernel_headers.h>
+#include <iptables/libiptc/ipt_kernel_headers.h>
 #include <linux/netfilter_ipv4/ip_tables.h>
 
 #ifdef __cplusplus
