diff -ur iptables-1.2.3-/extensions/libip6t_length.c iptables-1.2.3/extensions/libip6t_length.c
--- iptables-1.2.3-/extensions/libip6t_length.c	Sun Aug 12 13:16:49 2001
+++ iptables-1.2.3/extensions/libip6t_length.c	Mon Oct 15 15:40:04 2001
@@ -9,11 +9,6 @@
 #include <ip6tables.h>
 #include <linux/netfilter_ipv6/ip6t_length.h>
 
-struct ip6t_length_info {
-    u_int16_t	min, max;
-    u_int8_t	invert;
-};
-
 /* Function which prints out usage message. */
 static void
 help(void)
@@ -41,11 +36,9 @@
 static u_int16_t
 parse_length(const char *s)
 {
-	int len;
-	
-	len = string_to_number(s, 0, 0xFFFF);
+	unsigned int len;
 	
-	if (len == -1)
+	if (string_to_number(s, 0, 0xFFFF, &len) == -1)
 		exit_error(PARAMETER_PROBLEM, "length invalid: `%s'\n", s);
 	else
 		return (u_int16_t )len;
