--- iptables-1.4.8/extensions/libxt_owner.c.org	2010-06-07 21:59:55.086564655 +0200
+++ iptables-1.4.8/extensions/libxt_owner.c	2010-06-07 22:06:25.259858702 +0200
@@ -24,11 +24,15 @@
 	IPT_OWNER_PID   = 0x04,
 	IPT_OWNER_SID   = 0x08,
 	IPT_OWNER_COMM  = 0x10,
+	IPT_OWNER_NID   = 0x20,
+	IPT_OWNER_XID   = 0x40,
 	IP6T_OWNER_UID  = IPT_OWNER_UID,
 	IP6T_OWNER_GID  = IPT_OWNER_GID,
 	IP6T_OWNER_PID  = IPT_OWNER_PID,
 	IP6T_OWNER_SID  = IPT_OWNER_SID,
 	IP6T_OWNER_COMM = IPT_OWNER_COMM,
+	IP6T_OWNER_NID  = IPT_OWNER_NID,
+	IP6T_OWNER_XID  = IPT_OWNER_XID,
 };
 
 struct ipt_owner_info {
@@ -38,6 +42,8 @@
 	pid_t sid;
 	char comm[16];
 	u_int8_t match, invert;	/* flags */
+	u_int32_t nid;
+	u_int32_t xid;
 };
 
 struct ip6t_owner_info {
@@ -47,6 +53,8 @@
 	pid_t sid;
 	char comm[16];
 	u_int8_t match, invert;	/* flags */
+	u_int32_t nid;
+	u_int32_t xid;
 };
 
 /*
