diff -urN -x CVS -x .cvsignore iptables-1.2.2/Makefile userspace/Makefile
--- iptables-1.2.2/Makefile	Tue May  1 11:44:16 2001
+++ userspace/Makefile	Mon May 14 16:32:52 2001
@@ -91,24 +91,24 @@
 	cp $< $@
 
 ip6tables.o: ip6tables.c
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -c -o $@ $<
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -c -o $@ $<
 
 ip6tables: ip6tables-standalone.c ip6tables.o libiptc/libiptc.a
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
 
 $(DESTDIR)$(BINDIR)/ip6tables: ip6tables
 	@[ -d $(DESTDIR)$(BINDIR) ] || mkdir -p $(DESTDIR)$(BINDIR)
 	cp $< $@
 
 ip6tables-save: ip6tables-save.c ip6tables.o libiptc/libiptc.a
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
 
 $(DESTDIR)$(BINDIR)/ip6tables-save: ip6tables-save
 	@[ -d $(DESTDIR)$(BINDIR) ] || mkdir -p $(DESTDIR)$(BINDIR)
 	cp $< $@
 
 ip6tables-restore: ip6tables-restore.c ip6tables.o libiptc/libiptc.a
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
 
 $(DESTDIR)$(BINDIR)/ip6tables-restore: ip6tables-restore
 	@[ -d $(DESTDIR)$(BINDIR) ] || mkdir -p $(DESTDIR)$(BINDIR)
diff -urN -x CVS -x .cvsignore iptables-1.2.2/extensions/libipt_REJECT.c userspace/extensions/libipt_REJECT.c
--- iptables-1.2.2/extensions/libipt_REJECT.c	Mon Dec 18 06:09:52 2000
+++ userspace/extensions/libipt_REJECT.c	Mon May 14 16:32:55 2001
@@ -146,8 +146,13 @@
 {
 	const struct ipt_reject_info *reject
 		= (const struct ipt_reject_info *)target->data;
+	unsigned int i;
 
-	printf("--reject-with %s ", reject_table[reject->with].name);
+	for (i = 0; i < sizeof(reject_table)/sizeof(struct reject_names); i++)
+		if (reject_table[i].with == reject->with)
+			break;
+
+	printf("--reject-with %s ", reject_table[i].name);
 }
 
 struct iptables_target reject
diff -urN -x CVS -x .cvsignore iptables-1.2.2/ip6tables-save.c userspace/ip6tables-save.c
--- iptables-1.2.2/ip6tables-save.c	Tue Feb 27 10:59:48 2001
+++ userspace/ip6tables-save.c	Mon May 14 16:32:52 2001
@@ -51,7 +51,7 @@
 			if (iface[i] != '\0')
 				printf("%c", iface[i]);
 		} else {
-			if (iface[i] != '\0')
+			if (iface[i] == '\0')
 				printf("+");
 			break;
 		}
diff -urN -x CVS -x .cvsignore iptables-1.2.2/ip6tables.c userspace/ip6tables.c
--- iptables-1.2.2/ip6tables.c	Sat May  5 23:37:47 2001
+++ userspace/ip6tables.c	Mon May 14 16:32:52 2001
@@ -745,7 +745,7 @@
 		memset(mask + vialen + 1, 0, IFNAMSIZ - vialen - 1);
 	}
 	for (i = 0; vianame[i]; i++) {
-		if (!isalnum(vianame[i])) {
+		if (!isalnum(vianame[i]) && vianame[i] != '_') {
 			printf("Warning: wierd character in interface"
 			       " `%s' (No aliases, :, ! or *).\n",
 			       vianame);
diff -urN -x CVS -x .cvsignore iptables-1.2.2/iptables-save.c userspace/iptables-save.c
--- iptables-1.2.2/iptables-save.c	Thu Mar 15 16:12:02 2001
+++ userspace/iptables-save.c	Mon May 14 16:32:52 2001
@@ -48,7 +48,7 @@
 			if (iface[i] != '\0')
 				printf("%c", iface[i]);
 		} else {
-			if (iface[i] != '\0')
+			if (iface[i] == '\0')
 				printf("+");
 			break;
 		}
diff -urN -x CVS -x .cvsignore iptables-1.2.2/iptables.c userspace/iptables.c
--- iptables-1.2.2/iptables.c	Sun Mar 25 21:55:04 2001
+++ userspace/iptables.c	Mon May 14 16:32:52 2001
@@ -744,7 +744,7 @@
 		memset(mask + vialen + 1, 0, IFNAMSIZ - vialen - 1);
 	}
 	for (i = 0; vianame[i]; i++) {
-		if (!isalnum(vianame[i])) {
+		if (!isalnum(vianame[i]) && vianame[i] != '_') {
 			printf("Warning: wierd character in interface"
 			       " `%s' (No aliases, :, ! or *).\n",
 			       vianame);
diff -urN -x CVS -x .cvsignore iptables-1.2.2/patch-o-matic/irc-conntrack-nat.patch userspace/patch-o-matic/irc-conntrack-nat.patch
--- iptables-1.2.2/patch-o-matic/irc-conntrack-nat.patch	Thu May  3 02:30:11 2001
+++ userspace/patch-o-matic/irc-conntrack-nat.patch	Sun May 20 21:27:21 2001
@@ -260,8 +260,8 @@
 +					printk(KERN_WARNING
 +						"Forged DCC command from "
 +						"%u.%u.%u.%u: %u.%u.%u.%u:%u\n",
-+				HIPQUAD(ct->tuplehash[dir].tuple.src.ip),
-+						NIPQUAD(dcc_ip), dcc_port);
++				NIPQUAD(ct->tuplehash[dir].tuple.src.ip),
++						HIPQUAD(dcc_ip), dcc_port);
 +
 +				continue;
 +			}
