diff -ur iptables-1.4.12.2/extensions/libxt_pkttype.c iptables-1.4.12.2.new/extensions/libxt_pkttype.c
--- iptables-1.4.12.2/extensions/libxt_pkttype.c	2012-01-02 18:19:09.000000000 +0100
+++ iptables-1.4.12.2.new/extensions/libxt_pkttype.c	2012-02-23 09:58:15.914689222 +0100
@@ -7,6 +7,7 @@
 #include <stdio.h>
 #include <string.h>
 #include <xtables.h>
+#define __aligned_u64 __u64 __attribute__((aligned(8)))
 #include <linux/if_packet.h>
 #include <linux/netfilter/xt_pkttype.h>
 
