--- iptables-1.2.8_20030605/extensions/libipt_IPMARK.c.org	Sat May 31 17:31:31 2003
+++ iptables-1.2.8_20030605/extensions/libipt_IPMARK.c	Mon Nov  3 13:11:28 2003
@@ -33,7 +33,7 @@
 "  --and-mask value       logical AND ip address with this value becomes MARK\n"
 "  --or-mask value        logical OR ip address with this value becomes MARK\n"
 "\n",
-NETFILTER_VERSION);
+IPTABLES_VERSION);
 }
 
 static struct option opts[] = {
@@ -152,7 +152,7 @@
 struct iptables_target ipmark
 = { NULL,
     "IPMARK",
-    NETFILTER_VERSION,
+    IPTABLES_VERSION,
     IPT_ALIGN(sizeof(struct ipt_ipmark_target_info)),
     IPT_ALIGN(sizeof(struct ipt_ipmark_target_info)),
     &help,
