--- iptables-1.8.3/configure.ac.orig	2019-05-27 17:09:32.000000000 +0200
+++ iptables-1.8.3/configure.ac	2019-09-22 08:36:45.230554448 +0200
@@ -117,7 +117,7 @@
 AM_CONDITIONAL([ENABLE_CONNLABEL], [test "$enable_connlabel" = "yes"])
 
 if test "x$enable_bpfc" = "xyes" || test "x$enable_nfsynproxy" = "xyes"; then
-	AC_CHECK_LIB(pcap, pcap_compile,, AC_MSG_ERROR(missing libpcap library required by bpf compiler or nfsynproxy tool))
+	AC_CHECK_LIB(pcap, pcap_compile, [:], AC_MSG_ERROR(missing libpcap library required by bpf compiler or nfsynproxy tool))
 fi
 
 PKG_CHECK_MODULES([libnfnetlink], [libnfnetlink >= 1.0],
