--- iptables-1.6.0/iptables/xtables-compat-multi.c.orig	2015-12-09 13:55:06.000000000 +0100
+++ iptables-1.6.0/iptables/xtables-compat-multi.c	2016-02-26 21:42:01.268391511 +0100
@@ -28,6 +28,7 @@
 	{"ip6tables-compat-restore",	xtables_ip6_restore_main},
 	{"arptables",			xtables_arp_main},
 	{"arptables-compat",		xtables_arp_main},
+	{"ebtables",			xtables_eb_main},
 	{"ebtables-compat",		xtables_eb_main},
 	{NULL},
 };
