diff -urN iptables-1.1.2/Makefile iptables-1.1.2.new/Makefile
--- iptables-1.1.2/Makefile	Wed Oct 11 08:07:19 2000
+++ iptables-1.1.2.new/Makefile	Thu Oct 19 19:39:27 2000
@@ -60,24 +60,24 @@
 	cp $< $@
 
 ip6tables.o: ip6tables.c
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -c -o $@ $<
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -c -o $@ $<
 
 ip6tables: ip6tables-standalone.c ip6tables.o libiptc/libiptc.a
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
 
 $(DESTDIR)$(BINDIR)/ip6tables: ip6tables
 	@[ -d $(DESTDIR)$(BINDIR) ] || mkdir -p $(DESTDIR)$(BINDIR)
 	cp $< $@
 
 ip6tables-save: ip6tables-save.c ip6tables.o libiptc/libiptc.a
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
 
 $(DESTDIR)$(BINDIR)/ip6tables-save: ip6tables-save
 	@[ -d $(DESTDIR)$(BINDIR) ] || mkdir -p $(DESTDIR)$(BINDIR)
 	cp $< $@
 
 ip6tables-restore: ip6tables-restore.c ip6tables.o libiptc/libiptc.a
-	$(CC) $(CFLAGS) -DIPT_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
+	$(CC) $(CFLAGS) -DIP6T_LIB_DIR=\"$(IPT_LIBDIR)\" -rdynamic -o $@ $^ -ldl
 
 $(DESTDIR)$(BINDIR)/ip6tables-restore: ip6tables-restore
 	@[ -d $(DESTDIR)$(BINDIR) ] || mkdir -p $(DESTDIR)$(BINDIR)
diff -urN iptables-1.1.2/netfilter_ULOG-0.3/iptables/Makefile iptables-1.1.2.new/netfilter_ULOG-0.3/iptables/Makefile
--- iptables-1.1.2/netfilter_ULOG-0.3/iptables/Makefile	Thu Aug 10 13:48:23 2000
+++ iptables-1.1.2.new/netfilter_ULOG-0.3/iptables/Makefile	Thu Oct 19 19:40:33 2000
@@ -1,4 +1,6 @@
-CFLAGS = -DNETFILTER_VERSION=\"1.1.1\" -fPIC
+CFLAGS = $(COPT_FLAGS) -DNETFILTER_VERSION=\"1.1.2\" -fPIC -I../../include
+
+all: libipt_ULOG.so
 
 libipt_ULOG.so: libipt_ULOG_sh.o
 	ld -shared libipt_ULOG_sh.o -o libipt_ULOG.so
diff -urN iptables-1.1.2/netfilter_ULOG-0.3/libipulog/Makefile iptables-1.1.2.new/netfilter_ULOG-0.3/libipulog/Makefile
--- iptables-1.1.2/netfilter_ULOG-0.3/libipulog/Makefile	Thu Aug 10 13:51:58 2000
+++ iptables-1.1.2.new/netfilter_ULOG-0.3/libipulog/Makefile	Thu Oct 19 19:39:27 2000
@@ -1,5 +1,7 @@
 CC = gcc
-CFLAGS = -I./include # -g
+CFLAGS = $(COPT_FLAGS) -I./include # -g
+
+all: libipulog.a ulog_test
 
 ulog_test: ulog_test.c libipulog.a
 	$(CC) $(CFLAGS) -i ulog_test.c libipulog.a -o ulog_test
diff -urN iptables-1.1.2/netfilter_ULOG-0.3/ulogd/Makefile iptables-1.1.2.new/netfilter_ULOG-0.3/ulogd/Makefile
--- iptables-1.1.2/netfilter_ULOG-0.3/ulogd/Makefile	Mon Aug 14 10:28:23 2000
+++ iptables-1.1.2.new/netfilter_ULOG-0.3/ulogd/Makefile	Thu Oct 19 19:41:03 2000
@@ -8,7 +8,7 @@
 #  Normally You should not need to change anything below
 #
 CC = gcc
-CFLAGS = -I. -g -Wall $(INCIPULOG)
+CFLAGS = $(COPT_FLAGS) -I. -Wall $(INCIPULOG)
 SH_CFLAGS:=$(CFLAGS) -fPIC
 
 SHARED_LIBS+=$(foreach T,$(ULOGD_SL),extensions/ulogd_$(T).so)
diff -urN iptables-1.1.2/netfilter_ULOG-0.3/ulogd/extensions/ulogd_OPRINT.c iptables-1.1.2.new/netfilter_ULOG-0.3/ulogd/extensions/ulogd_OPRINT.c
--- iptables-1.1.2/netfilter_ULOG-0.3/ulogd/extensions/ulogd_OPRINT.c	Mon Aug 14 10:28:24 2000
+++ iptables-1.1.2.new/netfilter_ULOG-0.3/ulogd/extensions/ulogd_OPRINT.c	Thu Oct 19 19:39:27 2000
@@ -26,7 +26,7 @@
         ((unsigned char *)&addr)[1], \
         ((unsigned char *)&addr)[0]
 
-#define ULOGD_OPRINT_FILE	"/var/log/ulogd.pktlog"
+#define ULOGD_OPRINT_FILE	"/var/log/iptables.ulog"
 
 static FILE *of = NULL;
 
diff -urN iptables-1.1.2/netfilter_ULOG-0.3/ulogd/ulogd.c iptables-1.1.2.new/netfilter_ULOG-0.3/ulogd/ulogd.c
--- iptables-1.1.2/netfilter_ULOG-0.3/ulogd/ulogd.c	Mon Aug 14 10:28:24 2000
+++ iptables-1.1.2.new/netfilter_ULOG-0.3/ulogd/ulogd.c	Thu Oct 19 19:39:27 2000
@@ -19,17 +19,14 @@
 #include <libipulog/libipulog.h>
 #include "ulogd.h"
 
-#define MYBUFSIZ 2048
-
-
 #define DEBUGP ulogd_error
 
 #ifndef ULOGD_PLUGIN_DIR
-#define ULOGD_PLUGIN_DIR	"/usr/local/lib/ulogd"
+#define ULOGD_PLUGIN_DIR	"/usr/lib/iptables/ulogd"
 #endif
 
 #ifndef ULOGD_LOGFILE
-#define ULOGD_LOGFILE		"/var/log/ulogd.log"
+#define ULOGD_LOGFILE		"/var/log/ulogd"
 #endif
 
 #ifndef ULOGD_NLGROUP
@@ -213,7 +210,7 @@
 	load_plugins();	
 	
 	/* allocate a receive buffer */
-	buf = (unsigned char *) malloc(MYBUFSIZ);
+	buf = (unsigned char *) malloc(BUFSIZ);
 	
 	/* create ipulog handle */
 	h = ipulog_create_handle(ipulog_group2gmask(ULOGD_NLGROUP));
@@ -233,7 +233,7 @@
 		 * handle_packet */
 		while(1)
 		{
-			len = ipulog_read(h, buf, MYBUFSIZ, 1);
+			len = ipulog_read(h, buf, BUFSIZ, 1);
 			upkt = ipulog_get_packet(buf);	
 			DEBUGP("==> packet received\n");
 			handle_packet(upkt);
