--- iptables-1.4.12/libiptc/Makefile.am.wiget	2011-08-15 20:42:49.617576861 +0200
+++ iptables-1.4.12/libiptc/Makefile.am	2011-08-15 20:44:08.393921572 +0200
@@ -5,10 +5,7 @@ AM_CPPFLAGS      = ${regular_CPPFLAGS} -
 
 pkgconfig_DATA      = libiptc.pc
 
-lib_LTLIBRARIES     = libip4tc.la libip6tc.la libiptc.la
-libiptc_la_SOURCES  =
-libiptc_la_LIBADD   = libip4tc.la libip6tc.la
-libiptc_la_LDFLAGS  = -version-info 0:0:0 ${libiptc_LDFLAGS2}
+lib_LTLIBRARIES     = libip4tc.la libip6tc.la
 libip4tc_la_SOURCES = libip4tc.c
 libip4tc_la_LDFLAGS = -version-info 0:0:0
 libip6tc_la_SOURCES = libip6tc.c

--- iptables-1.4.16.2/libiptc/libiptc.ld.orig	1970-01-01 01:00:00.000000000 +0100
+++ iptables-1.4.16.2/libiptc/libiptc.ld	2012-10-09 17:47:42.606342693 +0200
@@ -0,0 +1,4 @@
+/* GNU ld script
+   Gather both libipt4c and libipt6c, using them only as needed */
+OUTPUT_FORMAT(elf32-i386)
+GROUP ( AS_NEEDED ( /usr/lib/libip4tc.so /usr/lib/libip6tc.so ) )
