--- iptables-1.4.12/libiptc/Makefile.am.wiget	2011-08-15 20:42:49.617576861 +0200
+++ iptables-1.4.12/libiptc/Makefile.am	2011-08-15 20:44:08.393921572 +0200
@@ -5,10 +5,7 @@ AM_CPPFLAGS      = ${regular_CPPFLAGS} -
 
 pkgconfig_DATA      = libiptc.pc
 
-lib_LTLIBRARIES     = libip4tc.la libip6tc.la libiptc.la
-libiptc_la_SOURCES  =
-libiptc_la_LIBADD   = libip4tc.la libip6tc.la
-libiptc_la_LDFLAGS  = -version-info 0:0:0 ${libiptc_LDFLAGS2}
+lib_LTLIBRARIES     = libip4tc.la libip6tc.la
 libip4tc_la_SOURCES = libip4tc.c
 libip4tc_la_LDFLAGS = -version-info 0:0:0
 libip6tc_la_SOURCES = libip6tc.c

